### PR TITLE
Tweak HalResource Performance

### PR DESCRIPTION
**Pull request by furinvader**
The HalResource's properties are now lazy - they transform child resources and links on demand only.
Also, it does not use `angular.copy` as it probably was the main source of the performance problem, that occurred when wrapping more than over 1000 child resources at once.

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://furinvader-feature-hal-resource-performance.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/' target='_blank'>Open stage</a> or <a href='http://furinvader-feature-hal-resource-performance.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com//teatro.log' target='_blank'>view logs</a>.

**Comment by ulferts** [comment]
@furinvader if you rebase onto/merge dev, some of the errors should disappear.

File: frontend/app/components/api/api-v3/hal-resources/work-package-resource.service.ts [hunk 1]
```diff
@@ -69,27 +69,27 @@ function wpResource(HalResource:typeof op.HalResource, NotificationsService:any,
       this.getForm()
         .catch(deferred.reject)
         .then(form => {
-          var plain_payload = form.payload.$source;
+          var plainPayload = form.payload.$plain();
```
Comment (oliverguenther) [inline frontend/app/components/api/api-v3/hal-resources/work-package-resource.service.ts:69]: This change breaks changing assignees. I verified when using $source again, the functionality was restored. What exactly is the difference between the two?

**Comment by oliverguenther** [comment]
LGTM with the exception of the mentioned issue.

File: frontend/app/components/api/api-v3/hal-resources/work-package-resource.service.ts (hunk 1 above)
Comment (furinvader) [inline frontend/app/components/api/api-v3/hal-resources/work-package-resource.service.ts:69]: I added this as a fix to one of the breaking specs.
I guess it is obsolete, as you have found the actual issues.

I'll remove the commit and push again.

**Pull request merged**